### PR TITLE
Create abstract and custom classes for particles

### LINKED ISSUE #447: Create custom particle subclass
As per discussion on #439 it'd be useful to have a method for generating custom particles as part of the `Particle` class. In particular, we want to be able to pass a mass and a charge to create custom dust grain particles.

### PR DESCRIPTION
**Pull request by namurphy**
My goals here are:

 - Create an abstract particle class
 - Have `Particle` inherit from that abstract class
 - Create a class for dimensionless particles that inherits from the abstract class
 - Create a class for custom (dimensional) particles that inherits from the abstract class

Thus far it looks like the classes will include `mass` and `charge` attributes.  I'm also wondering about creating a composite particle class, which would have things like the root mean square charge which comes up occasionally in `plasmapy.formulary`.

Closes #447.  Related to #687.  This pull request is brought to you by proudly procrastinating doing other things.

### CODE REVIEW

**Comment by StanczakDominik** [comment]
> This pull request is brought to you by proudly procrastinating doing other things.

For those about to postpone doing other work by writing some cool abstract classes, we'll salute you... later.

File: plasmapy/particles/particle_class.py [hunk 1]
```diff
@@ -1585,3 +1605,185 @@ def recombine(self, n: Integral = 1, inplace=False):
...
+        if isinstance(q, Real):
+            self._charge = q
+        elif isinstance(q, u.Quantity) and q.unit is u.dimensionless_unscaled:
+            self._charge = q.to(u.dimensionless_unscaled).value
+        elif q is None or q is np.nan:
+            self._charge = np.nan
+        else:
+            raise InvalidParticleError(
+                "The charge of a dimensionless particle must be a real number."
+            )
```
Comment (StanczakDominik) [inline plasmapy/particles/particle_class.py]: I'd have thought dumping non-dimensionless Astropy Quantities via `.si.value` or as multiple of electron charge would have been in the cards. Maybe that's not explicit enough, though.

File: plasmapy/particles/particle_class.py [hunk 2]
```diff
@@ -1585,3 +1605,185 @@ def recombine(self, n: Integral = 1, inplace=False):
...
+
+    # TODO: Add a root mean square charge attribute.
```
Comment (StanczakDominik) [inline plasmapy/particles/particle_class.py]: Okay, now I'm curious. What's that for?

File: plasmapy/particles/particle_class.py [hunk 3]
```diff
@@ -105,7 +110,23 @@ def _category_errmsg(particle, category: str) -> str:
...
+
+    @property
+    @abstractmethod
+    def mass(self) -> Union[u.Quantity, Real]:
+        pass
+
+    @property
+    @abstractmethod
+    def charge(self) -> Union[u.Quantity, Real]:
+        pass
```
Comment (StanczakDominik) [inline plasmapy/particles/particle_class.py]: I think I used `particle.name` in `RefactorTracker` (hope you don't mind I call it by branch name, it's simpler for me that way - I've been checking it out a lot lately). Should I switch to `str(particle)`?

File: plasmapy/particles/particle_class.py (hunk 2 above)
Comment (namurphy) [reply plasmapy/particles/particle_class.py]: Hm...I think that actually may be out of scope now.  The formulary includes some functions that have `z_mean` as a keyword argument in order to account for composite particles.  I had thought there were functions that allowed for `z_rms` to different from `z_mean` but I couldn't find them just now.  For the time being, I'd say we can leave `z_rms` out until we have a need for it.

**Comment by namurphy** [comment]
I think this is just about ready to merge.  Two thoughts:
 - I am wondering if I could replace some of the validation steps with `validate_quantity`.  One potential barrier to this is that the current implementation for a custom particle accepts real numbers for charge, in which case it's treated as the integer charge.
 - We can merge this after #757 so that we can do the rebase in this PR, since I reformatted the file with black which will probably create conflicts.

**Comment by namurphy** [comment]
- [ ] Add change log

File: plasmapy/particles/particle_class.py [hunk 4]
```diff
@@ -478,15 +499,18 @@ def __eq__(self, other) -> bool:
                 return self.particle == other_particle.particle
             except InvalidParticleError as exc:
                 raise InvalidParticleError(
-                    f"{other} is not a particle and cannot be "
-                    f"compared to {self}.") from exc
+                    f"{other} is not a particle and cannot be " f"compared to {self}."
```
Comment (rocco8773) [inline plasmapy/particles/particle_class.py]: make a single f-string

**Comment by namurphy** [comment]
As a side note, @StanczakDominik mentioned during code review that [hypothesis](https://hypothesis.readthedocs.io/en/latest/) may be a useful library for testing in `plasmapy.particles` and `plasmapy.formulary`

File: plasmapy/particles/particle_class.py [hunk 5]
```diff
@@ -98,14 +88,38 @@ def _category_errmsg(particle, category: str) -> str:
...
-class Particle:
+class AbstractParticle(ABC):
+    """
+    An abstract base class that defines the interface for particles.
+    """
+
+    @property
+    @abstractmethod
+    def mass(self) -> Union[u.Quantity, Real]:
+        pass
```
Comment (rocco8773) [inline plasmapy/particles/particle_class.py:107]:
```suggestion
        raise NotImplementedError
```

File: plasmapy/particles/particle_class.py [hunk 6]
```diff
@@ -98,14 +88,38 @@ def _category_errmsg(particle, category: str) -> str:
...
+
+    @property
+    @abstractmethod
+    def charge(self) -> Union[u.Quantity, Real]:
+        pass
```
Comment (rocco8773) [inline plasmapy/particles/particle_class.py:112]:
```suggestion
        raise NotImplementedError
```

**Pull request merged**